### ISSUE TITLE
Rewrite item/annotation links in notes when copying between libraries

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#2546]
https://forums.zotero.org/discussion/comment/405391/#Comment_405391

This is a bit rough.

Maybe you copy a parent item and all its child items together and we rewrite everything together, but maybe you copy a PDF first and then copy a standalone note later, or vice versa. (Eventually you might be able to copy a child note too, but I think currently it's not possible to do that independently.) If you copy a standalone note first…does it continue working for the person who added it, but no one else, or should we remove the links? If it keeps them, if you copy the item + PDF + annotations later, does it search all notes in the target library for `/users/[userID]/items` links that match the items being copied and update them? This seems somewhat unreasonable, so I'm inclined to just remove the links if the note isn't copied at the same time.

In any case, we can probably start by addressing the somewhat simpler case of copying a parent item with all its descendant items.

Will ideally to do this from the web library and iOS eventually too. If possible we should keep this modular so that the web library, at least, can reuse some of the logic here.

**Comment by AbeJellinek** [comment]
>If possible we should keep this modular so that the web library, at least, can reuse some of the logic here.

Where should it go? Is there currently any shared code between the client and the web library?

**Comment by dstillman** [comment]
I don't think so, though it's possible @tnajdek has copied some individual functions. The 'utilities' submodule? It seems like that might have some things in it that could be useful.

**Comment by tnajdek** [comment]
There are [bits](https://github.com/zotero/web-library/blob/6f9e6b74494d1d142ccc9cbffe4185168d999563/src/js/common/identifiers.js#L47) and [blobs](https://github.com/zotero/web-library/blob/c177755dab736ff659de0dfd9f8758dc4804a751/src/js/common/format.js#L9) (and few more) copied over from zotero/zotero but no actual shared code.

zotero/utilities seems to depend on some namespaces from Zotero (e.g. `Services`) so it wouldn't work for web library.

**Comment by dstillman** [comment]
Not really — it's meant to be usable in other environments, and it's currently used in translation-server (node) and the connector (browsers). There might just be a few things that either shouldn't be in there or should be rewritten with non-Mozilla-dependent code.

**Comment by dstillman** [comment]
(E.g., it looks like `cleanURL()`, which uses `Services`, is only used [in one recently added translator](https://github.com/zotero/translators/blob/3c1b1f2c6380a01a94bc9837da330b7f289debad/IDEA%20ALM.js#L154), which presumably doesn't work in the connector or translation-server.)

**Comment by tnajdek** [comment]
Happy to rewrite `cleanURL` to not use Mozilla-dependent code, probably just use [cleanURL](https://github.com/zotero/web-library/blob/c177755dab736ff659de0dfd9f8758dc4804a751/src/js/utils.js#L300) from web library, though I need to make sure if it's functionally identical.

Another slight problem with zotero/utils is how functions are declared in a way that prevents any tree-shaking, which would affect the size of the bundle if used in the web library. It maybe possible to fix this issue easily without breaking compatibility.
